master/internal/api_user.go

```diff
@@ -683,3 +683,116 @@ func (a *apiServer) PostUserActivity(
+		ctx, *curUser); err != nil {
+		return nil, status.Error(codes.PermissionDenied, err.Error())
+	}
+
+	tokenExpiration := time.Now().Add(user.TokenExpirationDuration)
+	if req.Lifespan != "" {
+		d, err := time.ParseDuration(req.Lifespan)
+		if err != nil || d < 0 {
+			return nil, status.Error(codes.InvalidArgument,
+				"Lifespan must be a Go-formatted duration string with a positive value")
+		}
+		tokenExpiration = time.Now().Add(d)
+	}
+
+	token, err := user.DeleteAndCreateLongLivedToken(
+		ctx, curUser.ID, user.WithTokenExpiresAt(&tokenExpiration))
+	if err != nil {
+		return nil, err
+	}
+	return &apiv1.PostLongLivedTokenResponse{LongLivedToken: token}, nil
+}
+
+func (a *apiServer) PostUserLongLivedToken(
+	ctx context.Context, req *apiv1.PostUserLongLivedTokenRequest,
+) (*apiv1.PostUserLongLivedTokenResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	targetFullUser, err := getFullModelUser(ctx, model.UserID(req.UserId))
+	if err != nil {
+		return nil, err
+	}
+	targetUser := targetFullUser.ToUser()
+	if err = user.AuthZProvider.Get().CanCreateUsersToken(ctx, *curUser, targetUser); err != nil {
+		return nil, status.Error(codes.PermissionDenied, err.Error())
+	}
+
+	tokenExpiration := time.Now().Add(user.TokenExpirationDuration)
+	if req.Lifespan != "" {
+		d, err := time.ParseDuration(req.Lifespan)
+		if err != nil || d < 0 {
+			return nil, status.Error(codes.InvalidArgument,
+				"Lifespan must be a Go-formatted duration string with a positive value")
+		}
+		tokenExpiration = time.Now().Add(d)
+	}
+
+	token, err := user.DeleteAndCreateLongLivedToken(
+		ctx, targetFullUser.ID, user.WithTokenExpiresAt(&tokenExpiration))
+	if err != nil {
+		return nil, err
+	}
+	return &apiv1.PostUserLongLivedTokenResponse{LongLivedToken: token}, nil
+}
+
+func (a *apiServer) GetLongLivedToken(
+	ctx context.Context, req *apiv1.GetLongLivedTokenRequest,
+) (*apiv1.GetLongLivedTokenResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if err = user.AuthZProvider.Get().CanCreateUsersOwnToken(ctx, *curUser); err != nil {
+		return nil, status.Error(codes.PermissionDenied, err.Error())
+	}
+
+	tokenInfo, err := user.GetLongLivedTokenInfo(ctx, curUser.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	var res *userv1.LongLivedTokenInfo = tokenInfo.Proto()
+
+	return &apiv1.GetLongLivedTokenResponse{LongLivedTokenInfo: res}, nil
+}
+
+func (a *apiServer) GetUserLongLivedToken(
+	ctx context.Context, req *apiv1.GetUserLongLivedTokenRequest,
+) (*apiv1.GetUserLongLivedTokenResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	targetFullUser, err := getFullModelUser(ctx, model.UserID(req.UserId))
+	if err != nil {
+		return nil, err
+	}
+	targetUser := targetFullUser.ToUser()
+	if err = user.AuthZProvider.Get().CanCreateUsersToken(ctx, *curUser, targetUser); err != nil {
```

Comment:
Same nit in the vein of separating View & Create permission for tokens; however, a stance could be made that if this was renamed to `CanAdministrateLongLivedToken` then this could be applicable for both cases.

Do you have a preference here on how these could be separated? Is there a case where we'd have an admin that can view other user token info, but not necessarily create tokens for other users?

Reply:
I can't think of an example but can separate out the View, Create and Delete permissions for tokens. It may come handy in future if permissions change for any type of user.

Currently, we are assuming that Admins will have Create/View/Delete token permissions for themselves and other users, whereas Non-Admins will only have Create/View/Delete token permissions for themselves.
Correct me if needed?

Reply:
I imagine there could be cases where a user with the role `Viewer` might not be capable of creating, viewing, or deleting even tokens associated with their own account; where something like a Workspace Admin might only be able to see tokens belonging to users that are in their workspace. 

This is actually a good question for us to direct at product to get clarification for expectation on what each precanned role is expected to be able to do. But you are absolutely correct in the OSS/non-RBAC deployment case!

Reply:
I will have this documented as open questions to Product team. Thanks for clearing this out!